File: wrappers/dotnet/src/BbsSignatures/Models/ProofMessage.cs
namespace BbsSignatures
{
    public struct ProofMessage
    {
        /// <summary>
        /// The message
        /// </summary>
        public string Message;

        /// <summary>
        /// The proof type
        /// </summary>
        public ProofMessageType ProofType;

        /// <summary>
        /// Determines whether the specified <see cref="System.Object" />, is equal to this instance.
        /// </summary>
        /// <param name="obj">The <see cref="System.Object" /> to compare with this instance.</param>
        /// <returns>
        ///   <c>true</c> if the specified <see cref="System.Object" /> is equal to this instance; otherwise, <c>false</c>.
        /// </returns>
        /// <exception cref="System.NotImplementedException"></exception>
        public override bool Equals(object obj) => obj switch
        {
            ProofMessage msg => this == msg,
            null => false,
            _ => false
        };

        /// <summary>
        /// Converts to string.
        /// </summary>
        /// <returns>
        /// A <see cref="System.String" /> that represents this instance.
        /// </returns>
        public override string ToString()
        {
            return $"{ProofType}: {Message}";
        }

        /// <summary>
        /// Returns a hash code for this instance.
        /// </summary>
        /// <returns>
        /// A hash code for this instance, suitable for use in hashing algorithms and data structures like a hash table. 
        /// </returns>
        public override int GetHashCode()
        {
            return $"{Message}{ProofType}".GetHashCode();
        }

        /// <summary>
        /// Implements the operator ==.
        /// </summary>
        /// <param name="left">The left.</param>
        /// <param name="right">The right.</param>
        /// <returns>
        /// The result of the operator.
        /// </returns>
        public static bool operator ==(ProofMessage left, ProofMessage right)
        {
            return left.Message == right.Message && left.ProofType == right.ProofType;
        }

        /// <summary>
        /// Implements the operator !=.
        /// </summary>
        /// <param name="left">The left.</param>
        /// <param name="right">The right.</param>
        /// <returns>
        /// The result of the operator.
        /// </returns>
        public static bool operator !=(ProofMessage left, ProofMessage right)
        {
            return left.Message != right.Message || left.ProofType != right.ProofType;
        }
    }
}

File: wrappers/python/ursa_bbs_signatures/models/ProofMessage.py
from enum import IntEnum, unique

@unique
class ProofMessageType(IntEnum):
    Revealed = 1
    HiddenProofSpecificBlinding = 2
    HiddenExternalBlinding = 3


class ProofMessage:
    def __init__(
        self,
        message: str,
        proof_type: ProofMessageType,
    ) -> None:
        """A class to hold proof messages

        Args:
            message: The message
            proof_type: The proof type
        """
        self.proof_type = proof_type
        self.message = message

    def __repr__(self) -> str:
        """Converts the object to a string representation
        Returns: a string representation
        """
        return f"{self.proof_type.name}: {self.message}"

    def __hash__(self) -> int:
        """Returns the hash value of self
        Returns: The hashed object
        """
        return hash(f"{self.message}{self.proof_type.value}")

    def __eq__(self, o: "ProofMessage") -> bool:
        """ Enables the usage of the == operator
        """
        return self.message == o.message and self.proof_type.value == o.proof_type.value

    def __ne__(self, o: object) -> bool:
        """ Enables the usage of the != operator
        """
        return not self.__eq__(o)